hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/SchemaOneDeletedBlocksTable.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.hadoop.hdds.utils.db.TableIterator;
+import org.apache.hadoop.hdds.utils.db.cache.CacheValue;
+import org.apache.hadoop.ozone.container.common.helpers.ChunkInfoList;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * For RocksDB instances written using DB schema version 1, all data is
+ * stored in the default column family. This differs from later schema
+ * versions, which put deleted blocks in a different column family.
+ * As a result, the block IDs used as keys for deleted blocks must be
+ * prefixed in schema version 1 so that they can be differentiated from
+ * regular blocks. However, these prefixes are not necessary in later schema
+ * versions, because the deleted blocks and regular blocks are in different
+ * column families.
+ * <p>
+ * Since clients must operate independently of the underlying schema version,
+ * This class is returned to clients using {@link DatanodeStoreSchemaOneImpl}
+ * instances, allowing them to access keys as if no prefix is
+ * required, while it adds the prefix when necessary.
+ * This means the client should omit the deleted prefix when putting and
+ * getting keys, regardless of the schema version.
+ * <p>
+ * Note that this class will only apply prefixes to keys as parameters,
+ * never as return types. This means that keys returned through iterators
+ * like {@link SchemaOneDeletedBlocksTable#getSequentialRangeKVs},
+ * {@link SchemaOneDeletedBlocksTable#getRangeKVs}, and
+ * {@link SchemaOneDeletedBlocksTable#iterator} will return keys prefixed
+ * with {@link SchemaOneDeletedBlocksTable#DELETED_KEY_PREFIX}.
+ */
+public class SchemaOneDeletedBlocksTable implements Table<String,
+        ChunkInfoList> {
+  public static final String DELETED_KEY_PREFIX = "#deleted#";
+
+  private final Table<String, ChunkInfoList> table;
+
+  public SchemaOneDeletedBlocksTable(Table<String, ChunkInfoList> table) {
+    this.table = table;
+  }
+
+  @Override
+  public void put(String key, ChunkInfoList value) throws IOException {
+    table.put(prefix(key), value);
+  }
+
+  @Override
+  public void putWithBatch(BatchOperation batch, String key,
+                           ChunkInfoList value)
+          throws IOException {
+    table.putWithBatch(batch, prefix(key), value);
+  }
+
+  @Override
+  public boolean isEmpty() throws IOException {
+    return table.isEmpty();
+  }
+
+  @Override
+  public void delete(String key) throws IOException {
+    table.delete(prefix(key));
+  }
+
+  @Override
+  public void deleteWithBatch(BatchOperation batch, String key)
+          throws IOException {
+    table.deleteWithBatch(batch, prefix(key));
+  }
+
+  /**
+   * Because the actual underlying table in this schema version is the
+   * default table where all keys are stored, this method will iterate
+   * through all keys in the database.
+   */
+  @Override
+  public TableIterator<String, ? extends KeyValue<String, ChunkInfoList>>
+      iterator() {
+    return table.iterator();
+  }
```

Comment:
Throw UnsupportedOperation for this implementation and schema 2 implementation.

Reply:
In order to not support this method, the `KeyValueBlockIterator` will need a different way to gain access to the iterator it uses internally when filtering blocks by prefix. The updated version of the code will make this class internal to the `AbstractDatanodeStore`, and callers can get it using getters in the `DatanodeStore` interface that return the `KeyValueBlockIterator`'s interface: `BlockIterator`. This way, the `AbstractDatanodeStore` can initialize it with an iterator retrieved from the block data table before wrapping it in a class to disable access to this method.